File: Lib/src/cf/juh9870/lists/DoublyLinkedList.java
package cf.juh9870.lists;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class DoublyLinkedList<T> implements Iterable<T> {
	protected Node<T> last;
	private Node<T> first;
	private int length;

	public DoublyLinkedList() {
	}

	public DoublyLinkedList(Iterable<T> source) {
		addAll(source);
	}

	public DoublyLinkedList(T... elements) {
		addAll(Arrays.asList(elements));
	}

	protected Node<T> getFirst() {
		return first;
	}

	public Node<T> getLast() {
		return last;
	}

	public int getLength() {
		return length;
	}

	@Override
	public Iterator<T> iterator() {
		return new Iterator<>(this, false);
	}

	public Iterator<T> reverseIterator() {
		return new Iterator<>(this, true);
	}

	public Iterable<Node<T>> nodes() {
		return nodes(false);
	}

	public Iterable<Node<T>> nodes(boolean reverse) {
		Node<T> node = reverse ? last : first;
		List<Node<T>> result = new ArrayList<>();
		while (node != null) {
			result.add(node);
			node = reverse ? node.prev : node.next;
		}
		return result;
	}

	private Node<T> getNode(int index) {
		return getNode(index, false);
	}

	private Node<T> getNode(int index, boolean reverse) {
		if (index < 0 || index >= length)
			throw new IndexOutOfBoundsException("Index " + index + " is out of [0," + (length - 1) + "] range");
		Node<T> node = reverse ? last : first;
		while (index-- > 0) node = reverse ? node.prev : node.next;
		return node;
	}

	public T get(int index) {
		return getNode(index).value;
	}

	public void set(T value, int index) {
		getNode(index).value = value;
	}

	public void add(T element) {
		last = new Node<T>(element, last, null);
		if (first == null) first = last;
		length++;
	}

	public void addAll(Iterable<T> source) {
		for (T e : source) add(e);
	}

	public void delete(int index) {
		Node<T> node = getNode(index);
		if (node.hasNext())
			node.next.prev = node.prev;
		else
			last = node.prev;

		if (node.hasPrev())
			node.prev.next = node.next;
		else
			first = node.next;

		length--;
	}

	public DoublyLinkedList<T> fromLasts(int amount) {
		if (amount < 1 || amount > length)
			throw new IllegalArgumentException("Amount " + amount + " is out of [1," + (length - 1) + "] range");
		Iterator<T> enumerator = reverseIterator();
		DoublyLinkedList<T> list = new DoublyLinkedList<T>();

		while (amount-- > 0) {
			list.add(enumerator.next());
		}

		return list;
	}

	static class Node<T> {
		public T value;
		public Node<T> prev;
		public Node<T> next;

		public Node(T value, Node<T> prev,
					Node<T> next) {
			if (value == null) throw new IllegalArgumentException("Value must be non null");
			this.value = value;
			this.prev = prev;
			if (this.prev != null) this.prev.next = this;
			this.next = next;
			if (this.next != null) this.next.prev = this;
		}

		public boolean hasNext() {
			return next != null;
		}

		public boolean hasPrev() {
			return prev != null;
		}
	}

	class Iterator<T> implements java.util.Iterator<T> {
		Node<T> cur;
		boolean reverse;

		private Iterator(DoublyLinkedList<T> list, boolean reverse) {
			this.reverse = reverse;
			if (reverse) cur = list.last;
			else cur = list.first;
		}

		@Override
		public boolean hasNext() {
			return cur != null;
		}

		@Override
		public T next() {
			T data = cur.value;
			cur = reverse ? cur.prev : cur.next;
			return data;
		}
	}
}

File: JuhLib/List/DoublyLinkedList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace Juh.List
{
    public class DoublyLinkedList<T> : IEnumerable<T>
    {
        public DoublyLinkedList()
        {
        }

        public DoublyLinkedList(IEnumerable<T> source) : this()
        {
            AddAll(source);
        }

        public DoublyLinkedList(params T[] elements) : this()
        {
            AddAll(elements);
        }

        public DoublyLinkedListNode<T> First { get; protected set; }
        public DoublyLinkedListNode<T> Last { get; protected set; }

        public int Length { get; private set; }

        public IEnumerator<T> GetEnumerator()
        {
            return (from node in Nodes() select node.Value).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public IEnumerator<T> GetReverseEnumerator()
        {
            return (from node in Nodes(true) select node.Value).GetEnumerator();
        }

        public IEnumerable<DoublyLinkedListNode<T>> Nodes(bool reverse = false)
        {
            var node = reverse ? Last : First;
            while (node != null)
            {
                yield return node;
                node = reverse ? node.Prev : node.Next;
            }
        }

        private DoublyLinkedListNode<T> GetNode(int index, bool reverse = false)
        {
            if (index < 0 || index >= Length)
                throw new IndexOutOfRangeException($"Index {index} is out of [0,{Length - 1}] range");
            var node = reverse ? Last : First;
            while (index-- > 0) node = reverse ? node.Prev : node.Next;
            return node;
        }

        public T Get(int index)
        {
            return GetNode(index).Value;
        }

        public void Set(T value, int index)
        {
            GetNode(index).Value = value;
        }

        public void Add(T element)
        {
            Last = new DoublyLinkedListNode<T>(element, Last);
            if (First == null) First = Last;
            Length++;
        }

        public void AddAll(IEnumerable<T> source)
        {
            foreach (var e in source) Add(e);
        }

        public void Delete(int index)
        {
            var node = GetNode(index);
            if (node.HasNext())
                node.Next.Prev = node.Prev;
            else
                Last = node.Prev;

            if (node.HasPrev())
                node.Prev.Next = node.Next;
            else
                First = node.Next;

            Length--;
        }

        public DoublyLinkedList<T> FromLasts(int amount)
        {
            if (amount < 1 || amount > Length)
                throw new ArgumentOutOfRangeException(nameof(amount), $"{amount} is out of [1,{Length - 1}] range");
            var enumerator = GetReverseEnumerator();
            var list = new DoublyLinkedList<T>();

            while (amount-- > 0)
            {
                enumerator.MoveNext();
                list.Add(enumerator.Current);
            }

            return list;
        }
    }

    public class DoublyLinkedListNode<T>
    {
        public DoublyLinkedListNode(T value, DoublyLinkedListNode<T> prev = null,
            DoublyLinkedListNode<T> next = null)
        {
            Value = value ?? throw new ArgumentNullException(nameof(value));
            Prev = prev;
            if (Prev != null) Prev.Next = this;
            Next = next;
            if (Next != null) Next.Prev = this;
        }

        public T Value { get; set; }
        public DoublyLinkedListNode<T> Prev { get; set; }
        public DoublyLinkedListNode<T> Next { get; set; }

        public bool HasNext()
        {
            return Next != null;
        }

        public bool HasPrev()
        {
            return Prev != null;
        }
    }
}